### ISSUE TITLE
Memory hit when generating swagger

### ISSUE THREAD

**Issue by statler** [RicoSuter/NSwag#3306]
I have an API which generates a relatively large swagger document (2.1mb). This is because there are a lot of models and DTOs, and heaps of controllers and methods. When I generate the swagger on startup, it adds 300Mb to process memory which is never collected. I am trying to understand why this occurs and what I can do about it.  Without this line:

services.AddSwaggerDocument()

my memory footprint on startup is 270Mb. When I uncomment it, memory increases to 600Mb.

Any insights as to why would be appreciated. At the moment I simply have to turn off swagger generation in production, which I would prefer not to do obviously. 

I would have thought that swagger generation simply created and delivered a static file? What else is going on to explain this - is there some kind of .net core caching that is triggered by reflection or something. I am truly stumped.

**Comment by jeremyVignelles** [comment]
You could generate a static swagger.json during development and serve it at runtime, but you would need to make sure that it is always up to date. Have a look at NSwag.MSBuild. You would still need the NSwag library in your project for the document generation.

As for the high memory usage, several things might explain that:
-ASP.net core API explorer : don't know if they keep a cache of the generated models
- NSwag : the generated document is likely kept into memory
- Namotion.Reflection : this is a library that is used by NSwag to know whether a property is nullable and to get the documentation on each property, method and class. It does keep reflected types in a cache that you can clear by using `CachedType.ClearCache()`

**Comment by RicoSuter** [comment]
In theory we could automatically clear the reflection cache after the documents have been generated.

**Comment by jeremyVignelles** [comment]
Agreed, since that's just a cache, I think that it would be a good practice because we don't need it anymore.

**Comment by statler** [comment]
Thankyou for the information:

RicoSuter - can you advise how that would be done?

jeremyVignelles - my memory analysis shows that Namotion.Reflection is definitely contributing, but does not seem to be enough on its own.

**Comment by jeremyVignelles** [comment]
What Rico suggest is just calling `CachedType.ClearCache()` at the end of the generation process. It would free the memory used by Namotion.Reflection.

Do you know which other part of the code is contributing the most to the memory usage ?

**Comment by RicoSuter** [comment]
> Do you know which other part of the code is contributing the most to the memory usage ?

That would be interesting, indeed. I would have expected reflection as all other stuff can probably be GCed...

**Comment by statler** [comment]
OK - so hopefully this is useful?

The first is a screenshot of the full heap straight after loading the document. The second is the diff


![AfterSwaggerMemDump](https://user-images.githubusercontent.com/337991/108184688-96287e00-715f-11eb-8240-4c0b194c3873.png)

![MemoryDiff](https://user-images.githubusercontent.com/337991/108184699-97f24180-715f-11eb-962b-8c59b4f30d23.png)

**Comment by statler** [comment]
With regards the 3 options:
1. API Explorer
2. NSWAG
3. Namotion.Reflection 

None of these appear particularly useful once the swagger is generated. How can I get rid of them post-delivery?

**Comment by jeremyVignelles** [comment]
Start with clearing Namotion cache, that should get rid of the top 2.

Don't know about the others, but I don't know if you can clear the API explorer cache. That would be a question for the asp.net core team, not NSwag, unless NSwag is keeping track of the API explorer entities.

**Comment by RicoSuter** [comment]
I think NSwag only keeps the raw JSON (see Tuple<String, Exception, DateTimeOffset>) but not an OpenApiDocument or API Explorer stuff..

**Comment by statler** [comment]
Thanks for the advice on CachedType.ClearCache()

Sorry to be obtuse, but I really don't understand where I would do this - is it somewhere in AddSwaggerDocument?

                services.AddSwaggerDocument(document =>
                {
                }

**Comment by RicoSuter** [comment]
I think it should be in: 

```
app.UseOpenApi(c => c.PostProcess += { CachedType.ClearCache(); });
```

**Comment by statler** [comment]
Thank you. I have tried this and unfortunately it has very little effect. Does that mean it has to be API explorer?

**Comment by jeremyVignelles** [comment]
Please send an updated screenshot of your investigations. Also make sure that GC has been triggered before making conclusions.

You could also force a `GC.Collect()` after clearing the cache.

**Comment by RicoSuter** [comment]
But you see that the Namotion.Reflection stuff is gone?

**Comment by statler** [comment]
Yeah - but there is a heap of Nswag and NSchema stuff

![After GC](https://user-images.githubusercontent.com/337991/108190770-7d6f9680-7166-11eb-8b18-bccbd6e0fa20.png)


            app.UseOpenApi(a =>
            {
                a.PostProcess = (document, _) =>
                {
                    document.Schemes = new[] { OpenApiSchema.Https, OpenApiSchema.Http };
                    CachedType.ClearCache();
                    GC.Collect();
                };
            });

**Comment by RicoSuter** [comment]
Ok, with

    //XmlDocs.ClearCache();
    //CachedType.ClearCache();
    //GC.Collect();

![image](https://user-images.githubusercontent.com/2603405/108246885-41323b00-7152-11eb-91a4-89a686b02d42.png)

With 

    //XmlDocs.ClearCache();
    //CachedType.ClearCache();
    GC.Collect();

![image](https://user-images.githubusercontent.com/2603405/108246919-4e4f2a00-7152-11eb-9170-f16e87424afe.png)

and with

    XmlDocs.ClearCache();
    CachedType.ClearCache();
    GC.Collect();

![image](https://user-images.githubusercontent.com/2603405/108246995-61fa9080-7152-11eb-90b3-c5505ddaaae6.png)

Still needs more investigation...

Testing with NSwag.Sample.NET50 and this change in the middleware

![image](https://user-images.githubusercontent.com/2603405/108247158-95d5b600-7152-11eb-8ebb-5c9994d14c07.png)

**Comment by statler** [comment]
Thanks for the investigation Rico. At the moment I am having to prevent document generation in production which achieves the result, but isnt a long term solution as the interface will be quite import.

Is there any option for clearing these caches out without changing middleware?

**Comment by RicoSuter** [comment]
As you can see here, you can clear the caches in PostProcess (as already said). 

![image](https://user-images.githubusercontent.com/2603405/108324804-aaed2c00-71c8-11eb-99f9-fc4319c17a35.png)

**Comment by statler** [comment]
Ah - sorry mate - misunderstood. I thought those things were tied up in the internals somehow.

**Comment by RicoSuter** [comment]
No worries. At the moment the XmlDocs.ClearCache() and CachedType.ClearCache() is not really in the middleware (I added it for testing, screenshot is not actual code).

**Comment by statler** [comment]
Thanks - got it now. Unfortunately not getting any results though. I made a quick vid of me running it with the ClearCache, which does not seem to give me any joy - and then with the addswagger commented out. Does this leave me back at ApiExplorer as the culprit? The previously screenshotted NSwag resources are still present in the memory heap aft gc - though the NJsonSchema objects are gone.

https://user-images.githubusercontent.com/337991/108326576-8328b300-721e-11eb-8a2c-aafe7a404709.mp4

**Comment by RicoSuter** [comment]
I tried to only inject the API Explorer stuff into an operation to see whether it leads to problems, and I did not see the same (big) increase as with NSwag... (at least in Task Explorer, but this is very shaky to use of course).

**Comment by statler** [comment]
Your screenshots seemed to show that you managed to get it back down though didn't they? Or am I reading that wrong. No matter what I do to the Cache or GC, it stays about 300Mb higher after generating - the methods are definitely being hit, and they seem to deal with the Namotion and NSchema objects, but that doesn't make a lot of difference. The NSwag objects remain, but they aren't big enough to account for the 300mb increase

**Comment by RicoSuter** [comment]
It's strange, did a Memory Perf session: 

![image](https://user-images.githubusercontent.com/2603405/108330941-8300c700-71ce-11eb-9bfe-0cfbcae14dbd.png)

As you can see the arrow shows the clean cache thing, so no real diff...

**Comment by statler** [comment]
OK - I think the summary of where we are currently at:
- We know that AddSwagger adds (possibly significant) memory consumption. With a big document I have got to 300mb
- Whatever the references are they would appear to be static, or references held, as they arent affected by GC
- They do not appear to be XMLDocs cache or Types cache
- There are some NSwag objects hanging about, but they do not account for the size of the memory increase observed

Is this one you just want me to leave with you? Or do you think I need to approach the .Net Core team to get them in on the matter? Is there anything useful I can contribute at this point?

Thanks heaps for your really prompt help

**Comment by RicoSuter** [comment]
Q: If you look at your original screenshot, you see eg that RequestDelegate is 7kb but it actually has allocated 130mb, right? Isnt this actually a problem with IIS memory mgmt?

**Comment by SebastianStehle** [comment]
I would download a trial of dotMemory, it provides better statistics.

**Comment by statler** [comment]
After the cache clear and GC, the additional memory usage is all unmanaged memory. The managed memory seems to be behaving as expected - I expect the framework to cost some unallocated memory, but a 300Mb blowout to load a 2Mb JSON seems a bit excessive. Basically, 15s-22s is the doc generation. Something is allocating a stack of unmanaged process in the AddSwagger method, and I don't really know how to figure out what it is.

With swagger

![image](https://user-images.githubusercontent.com/337991/108341915-4b296c00-722e-11eb-93f4-bd045f04f613.png)

Without swagger

![image](https://user-images.githubusercontent.com/337991/108342498-fcc89d00-722e-11eb-8c65-2ea5b6cec8b8.png)

**Comment by SebastianStehle** [comment]
I think it is just the  CLR. Unmanaged memory are also memory pools that are allocated from the runtime and never given back, until there is memory pressure from the OS. It is a normal behavior.

But of course you can improve it by reducing the allocations. I think the 50MB in heap 2 are more interesting, because this are just the caches and so on that are probably not needed. I really like these charts, that you get when you open a snapshot: https://github.com/ChilliCream/hotchocolate/issues/3006

You should collect Snapshots because then dotMemory also triggers the GC.

**Comment by SebastianStehle** [comment]
I am interested in this topic as well and in my case a lot of allocations come from xml comments:

![image](https://user-images.githubusercontent.com/1236435/109061309-cb484980-76e6-11eb-89d9-f126435219ad.png)

Especially xpaths seem to be very resource hungry. Could be worth to optimize this and replace it with simple XmlReader.

**Comment by lahma** [comment]
I've been [optimizing NSwag a bit](https://twitter.com/markolahma/status/1375542313859096578) as I got somewhat frustrated to MSBuild task execution duration. So far I haven't investigated neither "after" memory usage nor XML doc handling performance. I'm running this against an API that hasn't got any XML comments. Does anyone have good larger than trivial publicly available source tree which has this generation enabled and that would allow some profiling?

**Comment by SebastianStehle** [comment]
You can check out https://github.com/Squidex/squidex

**Comment by lahma** [comment]
@SebastianStehle I wasn't able to quite to identify the mysterious memory usage (wen up and then down after GCs), but I created a PR to speed up and reduce memory allocations in documentation generation though https://github.com/RicoSuter/Namotion.Reflection/pull/71 .

**Comment by SebastianStehle** [comment]
I was not talking about high usage, only about allocations

**Comment by lahma** [comment]
Sorry, got confused with OP, but nonetheless, hopefully helps 🙂

**Comment by SebastianStehle** [comment]
Yes, thank you very much

**Comment by lahma** [comment]
@SebastianStehle I see that there was a new release of NSwag just now and it contains the tweaks I made to Namotion.Reflection. Maybe you can give it a try and report back?

**Comment by SebastianStehle** [comment]
Sure, I will...

**Comment by SebastianStehle** [comment]
Looks very good...

**Comment by Havunen** [comment]
It seems this cache does not get cleared ? 

https://github.com/RicoSuter/Namotion.Reflection/blob/b0caf26383a3b49aba1a1a24c290b7543db95920/src/Namotion.Reflection/Performance/CachingXDocument.cs#L12

![image](https://user-images.githubusercontent.com/2021355/151833775-6fc9f71c-e3bf-4669-9a52-a78f17ddbe14.png)


![image](https://user-images.githubusercontent.com/2021355/151833791-b3972ed6-270d-4585-9a50-91a5aad63ef4.png)

**Comment by jeremyVignelles** [comment]
> * It does keep reflected types in a cache that you can clear by using `CachedType.ClearCache()`

@Havunen : Have you tried that ?

**Comment by Havunen** [comment]
yeah
```
XmlDocs.ClearCache();
CachedType.ClearCache();
GC.Collect();
```



It did remove some Namotion.Reflection memory usage but a lot memory is still used.

![image](https://user-images.githubusercontent.com/2021355/151841629-dc31d63a-48b5-4861-b287-155a0e7175cc.png)

**Comment by RicoSuter** [comment]
![image](https://user-images.githubusercontent.com/2603405/151843212-a70c2b28-2749-40db-8ce9-099d7b9d63fd.png)

So in the middleware we clear the cache (no GC.Collect though)... doesnt this help?
Any idea where the memory is allocated?

**Comment by lahma** [comment]
Is the memory rooted to Namotion, NJsonSchema or NSwag specifically? I think GC.Collect() might also be lazy at times if it sees no big need to do the cleanup. Are you offering the generation endpoint as HTTP API what triggers the usage? Based on the example you want the generation but want everything after that gone?

I usually generate the clients via separate MSBuild task so I don't see dangling memory issues as the generation isn't called normally at runtime.

**Comment by Havunen** [comment]
> Are you offering the generation endpoint as HTTP API what triggers the usage

yeah, after the open api 3 schema generation is done, we cache the JSON and just return that.

> So in the middleware we clear the cache (no GC.Collect though)... doesnt this help?
Any idea where the memory is allocated?

yeah it did help to some extent it removed most of Namotion.Reflection, but a lot memory is still used.


between the intial state and first api doc generation the by far biggest memory usage (+500MB) comes from NewtonSoft

![image](https://user-images.githubusercontent.com/2021355/151846492-512e2f98-be8c-4b0a-8e88-9814df3bfbe8.png)

Now it seems its dominated by Reflection from Newtonsoft? I'm still investigating this.
![image](https://user-images.githubusercontent.com/2021355/151846786-585787fc-67bb-4f4e-8d4b-01fd558f100c.png)


If I take another open api document from the different version of the same API then memory increases 100MB and again its Namotion.Reflection:

![image](https://user-images.githubusercontent.com/2021355/151847213-02a08aa1-4d64-45b4-bf95-21b1a279a1b7.png)

and yes I did call after every generation

```cs
                XmlDocs.ClearCache();
                CachedType.ClearCache();
```

**Comment by lahma** [comment]
I haven't done any profiling myself, but these fellas here:

https://github.com/RicoSuter/NSwag/blob/a423377944838cbe17f1b53e6c14069902689cb0/src/NSwag.Core/OpenApiDocument.Serialization.cs#L23-L27

Might hold to some memory. Might be worthwhile to check what happens if nulled with forceful reflection. I guess they are not needed after the generation if you are caching them. Maybe.

Anyway, each `DefaultContractResolver` instance (and its inheritors, like `PropertyRenameAndIgnoreSerializerContractResolver`) have their own cache, so every instance holds state.

**Comment by Havunen** [comment]
I have maybe found the problem:

It seems OpenApiRegistration is registered to IOC (ref https://github.com/RicoSuter/NSwag/blob/28b3fa42c3c8754ddb442e5e774ce60575931aff/src/NSwag.AspNetCore/Extensions/NSwagServiceCollectionExtensions.cs#L65-L86)  which holds reference to "Generator" (ref: https://github.com/RicoSuter/NSwag/blob/4f345e1adc9c49d8498d4018250760aa69dd8dd2/src/NSwag.AspNetCore/OpenApiDocumentRegistration.cs#L22)  which has reference to everywhere so GC is not able to clean stuff. 

![image](https://user-images.githubusercontent.com/2021355/151959617-bd4d97c2-c8d1-4e23-b72d-dd89d73bfcbc.png)

**Comment by RicoSuter** [comment]
If possible OpenApiDocumentRegistration should only receive the settings and create the generator on demand and then not hold it, right?

**Comment by RicoSuter** [comment]
v13.15.7

**Comment by Havunen** [comment]
There is still some System.Reflection types referenced from IoC -> settings -> ActualContractResolver, but its better than before
